.github/workflows/test-deploy.yaml

```diff
@@ -0,0 +1,128 @@
+        with:
+          # Workaround for a Github Checkout action bug
+          # https://github.com/actions/checkout/issues/1359#issuecomment-1567902034
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Set up Haskell
+        uses: ./.github/actions/setup-haskell
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+
+      - name: Install Wasp CLI
+        working-directory: waspc
+        run: ./run install
+
+      - name: Install Fly CLI
+        uses: superfly/flyctl-actions/setup-flyctl@v1
+
+      - name: Prepare server secrets for deployment
+        id: prepare_secrets
+        working-directory: waspc/examples/todoApp
+        run: |
+          # NOTE: This assumes env var values don't contain:
+          #   - The character `#`.
+          #   - An empty line (i.e., there are no multiline env values).
+          echo "ENV_VARS=$(
+            cat .env.server.headless \
+            | sed -E 's/#.*//' \
+            | awk 'NF {$1=$1;print}' \
+            | sed -E 's/^/--server-secret /' \
+            | paste -sd' ')" >> $GITHUB_ENV
+
+      - name: Deploy App to Fly.io
+        working-directory: waspc/examples/todoApp
+        run: |
+          # NOTE:The `yes` command is necessary because the `fly launch` command prompts for confirmation 
+          yes | wasp-cli deploy fly launch $APP_PREFIX $FLY_REGION --org wasp-testing ${{ env.ENV_VARS }}
+
+      - name: Save deployed app hostnames
+        id: save_hostnames
+        working-directory: waspc/examples/todoApp
+        run: |
+          echo "server_hostname=$(flyctl status -j -c fly-server.toml | jq -r '.Hostname')" >> "$GITHUB_OUTPUT"
+          echo "client_hostname=$(flyctl status -j -c fly-client.toml | jq -r '.Hostname')" >> "$GITHUB_OUTPUT"
+
+  smoke_test_app:
+    name: Run smoke tests on deployed app
+
+    runs-on: ubuntu-latest
+    needs: deploy_app_to_fly
+    env:
+      FLY_API_TOKEN: ${{ secrets.FLY_GITHUB_TESTING_TOKEN }}
+      SERVER_HOSTNAME: ${{ needs.deploy_app_to_fly.outputs.server_hostname }}
+      CLIENT_HOSTNAME: ${{ needs.deploy_app_to_fly.outputs.client_hostname }}
+
+    steps:
+      - name: Smoke test the server
+        run: |
+          curl --fail --silent -X POST https://$SERVER_HOSTNAME/operations/get-date
+
+      - name: Smoke test the client
+        run: |
+          curl --fail --silent https://$CLIENT_HOSTNAME
+
+  clean_up_deployed_fly_app:
+    runs-on: ubuntu-latest
+    name: Clean up deployed Fly app
+    needs: [generate_unique_app_prefix, deploy_app_to_fly]
+    # NOTE: Fly deployments can sometimes "fail" but still deploy the apps. We
+    # want to always clean them up.
+    if: needs.generate_unique_app_prefix.result == 'success'
+
+    env:
+      FLY_API_TOKEN: ${{ secrets.FLY_GITHUB_TESTING_TOKEN }}
+      APP_PREFIX: ${{ needs.generate_unique_app_prefix.outputs.app_prefix }}
+
+    steps:
+      - name: Install Fly CLI
+        uses: superfly/flyctl-actions/setup-flyctl@v1
```

Comment:
We want the clean up job to be the most robust, we don't want to end up with a bunch of leftovers in our Fly account. Therefore, I'd say that we should pin the Fly CLI version here (`with: version: blah`), so that changes in the CLI don't affect our ability to destroy apps.